src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,145 @@
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < INDEX_BY_ZERO) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You cannot add element by negative index");
+        }
+        if (array.length == size + INDEX_BY_ONE) {
+            array = grow();
+        }
+        if (index < array.length && index < size) {
+            copiedArray = (T[]) new Object[array.length];
+            System.arraycopy(array, INDEX_BY_ZERO, copiedArray, INDEX_BY_ZERO, index);
+            copiedArray[index] = value;
+            System.arraycopy(array, index, copiedArray, index + INDEX_BY_ONE, size + INDEX_BY_ONE);
+            System.arraycopy(copiedArray, INDEX_BY_ZERO, array, INDEX_BY_ZERO, size + INDEX_BY_ONE);
+            size++;
+            return;
+        } else if (index > size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You cannot add element by not existent index");
+        }
+        array[size] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < size && index >= INDEX_BY_ZERO) {
+            return array[index];
+        }
+        throw new ArrayListIndexOutOfBoundsException("There is not element by index: " + index);
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You cannot add element by not existent index");
+        } else if (index < INDEX_BY_ZERO) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You cannot add element by negative index");
+        }
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if ((index + INDEX_BY_ONE) == size) {
+            copiedArray = (T[]) new Object[array.length];
+            System.arraycopy(array, index, copiedArray, INDEX_BY_ZERO, size - index);
+            array[index] = null;
+            size--;
+            return copiedArray[INDEX_BY_ZERO];
+        }
+        if (index < size && index >= INDEX_BY_ZERO) {
+            copiedArray = (T[]) new Object[array.length];
+            System.arraycopy(array, index, copiedArray, INDEX_BY_ZERO, size - index);
+            System.arraycopy(copiedArray, INDEX_BY_ONE, array, index, size--);
+            return copiedArray[INDEX_BY_ZERO];
+        }
+        if (index < INDEX_BY_ZERO) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "You cannot remove element by negative index");
+        }
+        throw new ArrayListIndexOutOfBoundsException(
+                "You cannot remove element by not existent index");
```

Comment:
let's simplify this
this method can be written like this:
create and call `checkIndexInBounds` method
get T value you are about to remove
call System.arraycopy once
then        array[--size] = null;
and return value